### DISCUSSION TITLE
__FILE_NAME__ not defined in gcc < 12

### DISCUSSION THREAD

**Discussion by aeschbacher** [mesonbuild/meson#12477] [Q&A]
There are situations (when saving space is critical)  where preferring  the `__FILE_NAME__` [predefined macro](https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html) over `__FILE__` can make sense. Unfortunately the `__FILE_NAME__` macro is not available before gcc 12.

With GNU Make, it would be possible to `#define` it using `-D__FILE_NAME__` and some `$<` magic. I failed to find the right way of doing this with Meson build.  The obvious answer would be for sure to wait until it's possible to update to gcc 12 for the project (but upgrading a cross toolchain is a bigger change).

I'm wondering if it's possible in Meson with a reasonable effort  ?

**Comment by mitchgrout** [comment]
I don't think it's necessarily meson's place to provide such a macro. This could be readily resolved in your application without meson contributing at all, e.g.

```c
/* project/config.h */
#pragma once

#if !defined(__FILE_NAME__)
  static inline const char *basename(const char *path)
  { /* return the pointer after the last / in the string */ }

  // NOTE: Technically UB to define this name, but do what you need to
  #define __FILE_NAME__ basename(__FILE__)
#endif

**Reply by eli-schwartz** [reply]
I'm inclined to agree. The use of `__FILE__` is somewhat interesting since it can provide interesting information about where in the source tree the file was -- and using -fmacro-prefix-map, you can actually point to an on-disk location where the source file can be found in an auxiliary sources package. This is often used by distros to ship debug info in detached `foo-debug` packages.

Having just the basename of the file is much less interesting since you can technically statically code it. Just literally write `"foo.c"` everywhere you'd otherwise write `__FILE_NAME__`, and, admittedly, "remember to update that if you ever rename the file, rather than just move it to a new directory".

**Reply by mitchgrout** [reply]
I can see a use case where your source tree has unique filenames, e.g. your header structure may be along the lines of `<project/module>`, with the corresponding source being `project_module.c`. I've had thoughts around whether meson should provide some means of 'trimming' the existing `__FILE__` macro as per `-fmacro-prefix-map` but it's tricky to do right, and it practically destroys debuggability. In my case, I've opted for the same type of solution presented above; it's the least intrusive, and not particularly hard to implement.

**Reply by aeschbacher** [reply]
Thank you all for your answers. 

I realize that some context is probably missing. As I am building for an embedded device, my main concern is to reduce the  size of the generated binary. Using binutil's `string`, I can see following strings in the binary:

```
../../git/src/sub/dir/file.ccp
../../git/src/sub/dir/other.ccp
```

I don't care about the directory (there is no ambiguity in filenames).
I computed that by using `__FILE_NAME__` instead of `__FILE__` I could save multiple Kbytes by stripping the unneeded `../../git/src/sub/dir/` substrings. Using something like `#define __FILE_NAME__ basename(__FILE__)` does not help, as the whole strings would stay in the binary.

I just wanted to know if I missed some obvious solution, now I'll probably just wait until the cross-toolchain is updated to gcc 12

**Reply by eli-schwartz** [reply]
Does fmacro-prefix-map help here to chop off the leading portion of the macro expansion and replace it with `''`?

**Reply by aeschbacher** [reply]
Using `-fmacro-prefix-map` allows to easily remove the common part of the string (e.g. `../../git/src/`) , which is already a good improvement, thanks.

**Reply by aeschbacher** [reply]
I ended up with:

```meson
rel_source_dir = run_command([
    'realpath', '-m',
    '--relative-to=' + meson.project_build_root(),
    meson.project_source_root()
  ],
  check: false).stdout().strip()

flags = [  '-fmacro-prefix-map=' + rel_source_dir + '/=' ]

add_project_arguments(flags, language: ['c', 'cpp'])

```

**Reply by mitchgrout** [reply]
As a note as of meson 1.3.0 there is `fs.relative_to` which should be able to provide the same behaviour your `run_command`. I would be cautious about using `-fmacro-prefix-map` as while it does reduce your end binary size, I believe it can impact how tools like `gdb` are able to do instruction-to-source-line mapping

**Reply by aeschbacher** [reply]
Good to know, thanks
